README.md
# snowplow-micro-examples
![testing with Snowplow Micro](https://github.com/snowplow-incubator/snowplow-micro-examples/workflows/testing%20with%20Snowplow%20Micro/badge.svg)

Test that your website is tracking [Snowplow](https://snowplowanalytics.com/) events correctly.

## Quick start
```
# Launch the web app that we want to test:
docker-compose up -d web

# Launch snowplow micro
docker-compose up -d micro

# Run integration tests for snowplow tracking:
npm test
```
Did all tests pass? Then confidently push your web app to production!


Examples of how to apply [Snowplow Micro][snowplow-micro] to popular build/test strategies.

Current Snowplow Micro version: 1.2.1
If you have been using a previous version (v0.x) of Snowplow Micro in your test suites, you can also read the [Upgrading Micro](https://github.com/snowplow-incubator/snowplow-micro-examples#6-upgrading-micro) section.

## Table of Contents

[ Snowplow Micro](https://github.com/snowplow-incubator/snowplow-micro-examples#snowplow-micro)

[1. Local setup](https://github.com/snowplow-incubator/snowplow-micro-examples#1-local-setup)

[2. Github Actions](https://github.com/snowplow-incubator/snowplow-micro-examples#2-github-actions)

[3. Tracking design](https://github.com/snowplow-incubator/snowplow-micro-examples#3-tracking-design)

 - [3.1 Overview of the demo app](https://github.com/snowplow-incubator/snowplow-micro-examples#31-overview-of-the-demo-app)
 - [3.2 Event dictionary](https://github.com/snowplow-incubator/snowplow-micro-examples#32-event-dictionary)

[4. Testing with Snowplow Micro](https://github.com/snowplow-incubator/snowplow-micro-examples#4-testing-with-snowplow-micro)

 - [4.1 Snowplow Micro and Nightwatch](https://github.com/snowplow-incubator/snowplow-micro-examples#41-snowplow-micro-and-nightwatch)
 - [4.2 Snowplow Micro and Cypress](https://github.com/snowplow-incubator/snowplow-micro-examples#42-snowplow-micro-and-cypress)

[5. Additional resources](https://github.com/snowplow-incubator/snowplow-micro-examples#5-additional-resources)

[6. Upgrading Micro](https://github.com/snowplow-incubator/snowplow-micro-examples#6-upgrading-micro)


## Snowplow Micro

[Snowplow Micro][snowplow-micro-docs] is a very small version of a full [Snowplow](https://snowplowanalytics.com/) data collection pipeline that can be embedded in automated test suites and so enable Snowplow users to test their data collection setup and then release new versions of their apps, websites and servers with confidence that new changes won’t break the tracking set up. Snowplow Micro comes to fill a missing piece in automated test suites: the tests that validate that data collection has been setup properly.

To check data collection, it is necessary to fire a set of events, process them and see if the output of that processing is as expected. Snowplow Micro is easy to integrate with your test tool and then run as part of your automated testing workflow. This repository aims to show in detail all the steps to do this using Nightwatch and Cypress as examples of test tools, to build end-to-end GitHub Actions testing workflows.


## 1. Local setup

### 1.1 Prerequisites
We recommend setting up the following tools before starting:
 - Git
 - Docker and Docker-compose
 - Npm


### 1.2 Clone this repo, start Snowplow Micro and serve the app

```
$ git clone https://github.com/snowplow-incubator/snowplow-micro-examples.git

$ cd snowplow-micro-examples

$ docker-compose up
```
This will:

1. Start serving the app on localhost:8000
2. Launch Snowplow Micro, mounting the `micro` directory and setting the port 9090 for accessing the 4 endpoints: `/micro/all`, `/micro/good`, `/micro/bad` and `/micro/reset`.
   Inside the `micro` directory are:
    1. The [configuration for Snowplow Micro](./micro/micro.conf) and
    2. The [configuration for Iglu resolvers](./micro/iglu.json)


### 1.3 Install npm dependencies

```
$ npm install
```
This step will install Nightwatch and Cypress.


### 1.4 Run the tests

Our demo web app uses snowplow to track user activity. We will be testing that tracking is properly configured on our site.

To run all tests:
```
$ npm test
```

For running only [Nightwatch](https://nightwatchjs.org/) tests:
```
$ npm run test:nightwatch
```

For running only [Cypress](https://www.cypress.io/) tests:
```
# From the command line
$ npm run cypress:run

# To launch the Test Runner
$ npm run cypress:open
```

## 2. Github Actions
Inside the `.github/workflows/` directory you can find the `.yml` files we use to test this exaple app with Micro and Nightwatch/Cypress.
A general workflow file would definitely use the [Snowplow Micro][snowplow-micro-docs] step, which for our example is:

```
- name: Start Micro
    run: docker-compose up -d
    working-directory: snowplow-micro-examples
```
In order to use it, just make sure that:
1. Your `working-directory` contains the `micro/` directory that Micro will mount, containing a `micro.conf` and a `iglu.json` configuration for Micro and Iglu respectively.

2. Your `docker-compose.yml` file is adjusted accordingly

If you wanted to use `docker run` instead of `docker-compose`, the same step would be:
```
- name: Start Micro
    run: docker run --mount type=bind,source=$(pwd)/micro,destination=/config -p 9090:9090 snowplow/snowplow-micro:1.0.0 --collector-config /config/micro.conf --iglu /config/iglu.json &
    working-directory: snowplow-micro-examples
```


## 3. Tracking design

Our demo web app uses Snowplow to track user activity. Then, using Snowplow Micro we will be testing that tracking is properly configured on our site using, as examples, two populat web test tools, Nightwatch and Cypress.

### 3.1 Overview of the demo app

The example application is a simple ecommerce app consisting of just 3 pages:

 1. A start-up page with a sample login form (no authentication involved - see also the **Note** below)
 2. The shop page including a same-page cart
 3. The purchase-confirmation page

> **Note:**
> The form of the login-page is only for demonstrating the tracking of form events (which can also show that password fields are not being tracked). There is no authentication involved. While you can type anything and continue to the shop-page, we recommend that you do not use any personal data.

Each page serves the purpose of demonstrating possible event-tracking, which will then be tested using Snowplow Micro and a test tool.

### 3.2 Event dictionary

Using the [JavaScript Tracker][javascript-tracker]:
```
<!-- Snowplow starts plowing -->
<script type="text/javascript">
;(function(p,l,o,w,i,n,g){if(!p[i]){p.GlobalSnowplowNamespace=p.GlobalSnowplowNamespace||[];
p.GlobalSnowplowNamespace.push(i);p[i]=function(){(p[i].q=p[i].q||[]).push(arguments)
};p[i].q=p[i].q||[];n=l.createElement(o);g=l.getElementsByTagName(o)[0];n.async=1;
n.src=w;g.parentNode.insertBefore(n,g)}}(window, document, "script", "{% static 'ecommerce/js/sp.js' %}", "snowplow"));
```
The tracking implemented consists of:

1. [Pageview tracking][pageview-tracking]

    - This event happens when a user visits a page.
    - It is a predefined Snowplow event, that automatically captures the URL, referrer and page title.
    - Method call:
    ```
    window.snowplow('trackPageView');
    ```
    - Implemented in all 3 pages.

2. [Activity Tracking][activity-tracking]

    - This event happens when a user engages with a page (e.g. user scrolls).
    - It is a predefined Snowplow event, that automatically records the maximum scroll left-right, up-down in the last ping period.
    - Method call (before the `trackPageView` method call):
    ```
    window.snowplow('enableActivityTracking', {
        minimumVisitLength: number,
        heartbeatDelay: number
    });
    ```
    - Implemented in login-page and shop-page, with different `minimumVisitLength` and `heartBeat`:
    ```
    // login-page
    window.snowplow('enableActivityTracking', {
        minimumVisitLength: 20,
        heartbeatDelay: 20
    });

    // shop-page:
    window.snowplow('enableActivityTracking', {
        minimumVisitLength: 10,
        heartbeatDelay: 10
    });
    ```

3. [Form Tracking][form-tracking]
    - This set of events happens when a user interacts with a web form (e.g. focus on an form element, change a value of input-textarea-select element form, submit a form)
    - These are predefined Snowplow events that are of unstructured eventType and can be customized. Data captured:
        - **focus_form**: id, classes of the form and name, type, value of the form element that received focus.
        - **change_form**: name, type, new value of the element, id of the parent form
        - **submit_form**: id, classes of the form and name, type, value of all form elements
    - Implemented in the login-page:
    ```
    var opts = {
        forms: {
            denylist: []
        },
        fields: {
            denylist: ['user_password']
        }
    };

    window.snowplow('enableFormTracking', { options: opts });
    ```
    > **Note**:
    > By default, Form Tracking does not track password fields. We used the above `config` to demonstrate how you can ensure the Non-tracking of sensitive fields. With Snowplow Micro we can also later test that any blacklisting of forms is implemented correctly and that no sensitive fields are tracked.

4. [Tracking custom self-describing(unstructured) events][self-describing-tracking]

    1. cart-events ([schema](./iglu/schemas/test.example.iglu/cart_action_event/jsonschema/1-0-0))
        - These events happen when a user interacts with the cart, adding or removing items, using the Add-to-cart or Remove buttons.
        - This is a self-describing event that captures the type of cart interaction: "add" versus "remove".
        - We also want to add as [custom context][custom-context] the product involved in the cart-event, which is described by the product entity ([schema](./iglu/schemas/test.example.iglu/product_entity/jsonschema/1-0-0), see more below)
        - Implemented in the shop-page (see file [shoppage.js](./app/static/ecommerce/js/shoppage.js)):
        ```
        // TRACK cart_action_event (add)
        window.snowplow('trackSelfDescribingEvent', {
            event: {
                schema: 'iglu:test.example.iglu/cart_action_event/jsonschema/1-0-0',
                data: {
                    type: 'add',
                },
            },
            context: [
                {
                    schema: 'iglu:test.example.iglu/product_entity/jsonschema/1-0-0',
                    data: {
                        sku: sku,
                        name: title,
                        price: parseFloat(price),
                        quantity: parseInt(quantity)
                    }
                }
            ]
        });

        // TRACK cart_action_event (remove)
        window.snowplow('trackSelfDescribingEvent', {
            event: {
                schema: 'iglu:test.example.iglu/cart_action_event/jsonschema/1-0-0',
                data: {
                    type: 'remove',
                },
            },
            context: [
                {
                    schema: 'iglu:test.example.iglu/product_entity/jsonschema/1-0-0',
                    data: {
                        sku: sku,
                        name: title,
                        price: parseFloat(price),
                        quantity: parseInt(quantity)
                    }
                }
            ]
        });
        ```

    2. purchase-event ([schema](./iglu/schemas/test.example.iglu/purchase_event/jsonschema/1-0-0))
        - These events happen when a user completes the purchase of the products in their cart by clicking the Purchase button.
        - This is a self-describing event that captures the total amount of the transaction.
        - We also want to add as [custom contexts][custom-contexts] the products involved, each of which is described by the product entity ([schema](./iglu/schemas/test.example.iglu/product_entity/jsonschema/1-0-0), see more below)
        - Implemented in the shop-page (see file [shoppage.js](./app/static/ecommerce/js/shoppage.js)):
        ```
        // create the contexts array
        let productsContext = [];
        userCart.forEach(function(elt) {
            productsContext.push({
                schema: 'iglu:test.example.iglu/product_entity/jsonschema/1-0-0',
                data: {
                    sku: elt.itemSku,
                    name: elt.itemTitle,
                    price: parseFloat(elt.itemPrice),
                    quantity: parseInt(elt.itemQuant)
                }
            });
        });

        // TRACK purchase_event
        window.snowplow('trackSelfDescribingEvent', {
            event: {
                schema: 'iglu:test.example.iglu/purchase_event/jsonschema/1-0-0',
                data: {
                    total: parseFloat(total),
                },
            },
            context: productsContext
        });
        ```

5. Custom contexts

    - Predefined [webPage context][webpage-context] (enabled by default in JavaScript Tracker v3)
    - Product Entity ([schema](./iglu/schemas/test.example.iglu/product_entity/jsonschema/1-0-0)). This entity captures the data on a product involved in a cart or purchase event: sku, name, price, quantity.


## 4. Testing with Snowplow Micro

The purpose of this repo is to show how Snowplow Micro can be used to validate that the tracking implemented in a demo app is operating in the way we expect.

Simulate particular situations, and check that:
 - There are no bad events.
 - The data sent is as expected.
 - The right event data is sent with each event.
 - The right entities / contexts are attached to each event.
 - The right values are sent with each event.

In this repo you can also find information on how to set up trackers, how to make customised (unstructured) events, and, mainly, how to configure and add tests in your test tool of choice, using as examples the popular web test tools Nightwatch and Cypress, in order to demonstate the capabilities of Snowplow Micro.


#### The tests implemented in this repo are:

1) No bad events
2) Number of good events = number of expected good events
3) Ensuring that total number of events is right - expected number of good events + bad events (noBadEvents = True)
4) Checking proper values of structured and unstructured events are sent to Micro
5) Event With Property test - context, event type and schema
	- user puts in an event and we match by all 3 conditions (contexts, properties and schema - this determines whether or not the fake test event is equal to the one on micro - for both structured and unstructured)
6) Race condition test - to ensure that event x is always sent to Micro before event y (in our case we wanted to ensure cart action occurred before purchase)
7) Form tracking test to ensure blacklisted form fields are not tracked, and to ensure the right properties are sent for each field


### 4.1 Snowplow Micro and Nightwatch
Powered by Node.js, [Nightwatch.js](https://nightwatchjs.org/) is an open-source automated testing framework that aims at providing complete E2E (end to end) solutions to automate testing with Selenium Javascript for web-based applications, browser applications, and websites.

This framework relies on Selenium and provides several commands and assertions within the framework to perform operations on the DOM elements.

#### 4.1.1 How does Nightwatch JS work?
Nightwatch communicates over a restful API protocol that is defined by the W3C WebDriver API. It needs a restful HTTP API with a Selenium JavaScript WebDriver server.
In order to perform any operation i.e. either a command or assertion, Nightwatch usually requires sending a minimum of two requests.

It works as follows:

- The first request locates the required element with the given XPath expression or CSS selector.
- The second request takes the element and performs the actual operation of command or assertion.

#### 4.1.2 Getting Started

If you want to isolate Nightwatch JS testing, follow these steps, otherwise skip to [4.1.3 Simulating a User with Nightwatch](https://github.com/snowplow-incubator/snowplow-micro-examples#413-simulating-a-user-with-nightwatch).

Install nightwatch and a chrome driver which enables nightwatch to interact with the Chrome browser:

```
$ npm install nightwatch --save-dev
$ npm install chromedriver --save-dev
```

1. Create default json package
2. Create node modules folder and add nightwatch to dev dependencies
3. Add nightwatch.conf.js file (must contain basic configuration file)
4. Add a chrome driver binary for nightwatch configs - adds it into node modules - allows us to send commands to the chrome driver
5. Add `test:nightwatch` to the scripts section of the package.json
6. Write your own tests

Running nightwatch:
```
npm run test:nightwatch

```

#### 4.1.3 Simulating a User with Nightwatch
When using Nightwatch, one testing strategy is to simluate a user interaction with your app. This is useful so
that we can fire the exact events that a user would fire when using the app in the field.

In Nightwatch a test involves 3 phases:
1. Preparing a state:
    1. Reset Micro command - deleting the cache in Micro so that each test has independent events from other tests
    2. Configure Nightwatch to interact with your app
2. Taking an action:
    Nightwatch interacts with the app creating events
3. Making an assertion on the resulting state:
    Nighwatch requests the state of Micro and based on a test performs the corresponding assertions

#### 4.1.4 Organisation of tests

#### Commands

#### resetMicro
The resetMicro command can be added before each test as the beforeEach hook:
[Nightwatch Test Hooks](https://github.com/nightwatchjs/nightwatch-docs/blob/643e9a63a94deba6bd84bf2dd78cb27693620e20/guide/using-nightwatch/using-test-hooks.md)

*Example call*:

```
module.exports = {
    beforeEach: function(browser) {
        browser
            .resetMicro();
    },
    "test1" :function (browser0 {
        // code here
    }
}
```
*Arguments*: None

This command utilises the ability to reset Micro through the `/micro/reset` endpoint.

This is the general structure of how to create a command which aims to request information from Snowplow Micro using Nightwatch:

```
this.command = (callback) => {
        const request = require('request');

        request({
            url: 'http://localhost:9090/micro/all',
            json: true
        }, (err, res, body) => {
            if (err) {
                console.warn(error);
                return false;
            }
            callback(body);
        });
    };
```
#### Assertions

#### .noBadEvents

*Example call*:

```
    browser.assert.noBadEvents();
```
*Arguments*: None

This is arguably the most important assertion, if you only implement one this is a great place to start.
It ensures that all of your data is sent to Micro and ends up in good events, so all of your data is in your warehouse and you can interpret it as expected.

#### .noOfGoodEvents

*Example call*:

```
    browser.assert.noOfGoodEvents(2);
```
*Arguments*: Number of expected good events to be sent to Micro

An extension of noBadEvents, asserts that you sent the correct number of good events to Micro for a given event.
For example, you might expect the onClick() action to send 2 good events,
then you can make sure 2 are sent to good events and call noBadEvents on the same assertion as total number of events = number of good events + number of bad events.

#### .noOfTotalEvents

*Example call*:

```
    browser.assert.noOfTotalEvents(2);
```
*Arguments*: Number of expected total events sent to Micro

A further extension on noOfGoodEvents, this assertion ensures that both the correct number events are sent to Micro, and that no bad events are sent.
Using all 3 of these assertions consecutively provides the best assurance that every event you send is sent to your warehouse properly.

#### .orderOfEvents

*Example call*:

```
    browser.assert.orderOfEvents(events_list);

```
*Arguments*: events_list

Checks for when one event must be before the other so that your application works as expected.
This works by retrieving the derived timestamps for each event, and asserting that events fired in the specified order.
In our case we use this to test that a cart action occurs before purchasing an item.
If our application didn't get this order of events correct, then the application does not act as expected.
This can also be considered a race condition test.

#### .successfulEvent

*Example call*:

```
browser.assert.successfulEvent({
            "eventType": "unstruct",
            "schema": "iglu:test.example.iglu/cart_action_event/jsonschema/1-0-0",
            "values": {
                "type": "remove"
            },
            "contexts": [{
                "schema": "iglu:test.example.iglu/product_entity/jsonschema/1-0-0",
                "data": {
                    "name": "One-size summer hat",
                    "price": 15.5,
                    "quantity": 1
                }
            }]
        });
```
*Arguments*: test_event (with schema, eventType, values and context), number_of_occurences

Ensures that when an event is sent to Micro the correct parameters are sent as expected.
In our case we check that the schema, properties and contexts are correct:
 - We match what the user expects to be sent to what is received on Micro.
 - In the end, the number of matched events is retrieved and asserted to the expected number of occurrences.
 - By doing that, we can also specify which events we don't expect on Micro by setting the argument number_of_occurences=0.


### 4.2 Snowplow Micro and Cypress

[Cypress](https://www.cypress.io/) is an open [source](https://github.com/cypress-io/cypress) JavaScript End-to-End testing framework with extensive [documentation](https://docs.cypress.io/).
In this section we note few things that are specifically related to using this test tool with Snowplow Micro, describe the rationale behind the tests' organization used in this example and document the commands used.

#### 4.2.1 Introduction

Generally, a test involves 3 phases:
1. Prepare a state
2. Take an action
3. Assert on the resulting state

While Cypress considers as state the application's state, it is still a fact that an app is rarely an isolated system without side effects.

This is especially true when a tracking strategy is implemented, which means that any action can fire [events](https://docs.snowplowanalytics.com/docs/collecting-data/collecting-from-own-applications/snowplow-tracker-protocol/) through the [trackers](https://docs.snowplowanalytics.com/docs/collecting-data/collecting-from-own-applications/). There is an increasing number of reports (see Gartner) highlighting the importance of upstream data quality and Data Ops, which means that testing your Data Collection and trackers' implementation besides your product's features is of highest priority. Tracking is as important as your shipping, and that is why it is highly recommended that you include its testing in your E2E tests.

Cypress, even though it considers as [best practice](https://docs.cypress.io/guides/references/best-practices.html#Visiting-external-sites) to avoid requiring or testing 3rd party services, it still offers the ability to "talk" to 3rd party API's via [cy.request()](https://docs.cypress.io/api/commands/request.html), that:
 - Entirely bypasses CORS
 - Expects the server to exist and provide a response
 - Does not retry its assertions (as that could affect external state)
, which makes it great to use for querying Snowplow Micro's endpoints.
For example:
```
    cy.request({
        url: 'http://localhost:9090/micro/all',
        json:true
    });
```

So, following on the 3 test's phases:
1. Preparing state:
    1. Reset Micro
    2. Configure Cypress to visit your app
2. Actions:
    Cypress interacts with the app creating events
3. Assertions:
    Cypress sends requests to Micro, and attempts assertions on the responses



#### 4.2.2 Tests' organization

Another Cypress' recommendation for best [practices](https://docs.cypress.io/guides/references/best-practices.html#Having-tests-rely-on-the-state-of-previous-tests) is the decoupling of tests, which, for the case of testing with Snowplow Micro, would mean to run both the state-changing and the micro-requests in the same spec file.
However, there were some issues in doing so. More specifically, those issues had only to do with cases where links (or submit buttons) were clicked, in other words in cases where a window [unload event](https://developer.mozilla.org/en-US/docs/Web/API/Window/unload_event) was fired.

To describe the issue, we first describe what normally happens upon an unload event:
When a user clicks, for example, a link, on one hand the browser wants to navigate to the link, and on the other hand, the tracker (in our case the [Javascript Tracker][javascript-tracker]) tries to send the [link click][link-click-tracking] or the [submit form][form-tracking] events, while also storing them in local storage, just in case the events don't get sent before the page unloads.
While it is normal for browsers to cancel all requests, a cancelled request does not necessarily mean that the request did not reach the server, but that the client sending it, does not wait for an answer anymore. So, there is no way to know from client side whether the request (be it POST or GET) succeeded.

That problem was especially apparent when Micro was being queried in the same spec file with the app's actions. For example, POST requests appeared as cancelled in Cypress' test runner, but the events may have reached Micro.
Taking advantage of the fact that Cypress also clears browser cache when it changes spec file (see also this [issue](https://github.com/cypress-io/cypress/issues/686)), we decided to move the testing part of Micro into separate spec files.

The consequences of this decision are:
1. You need to ensure a naming strategy, and the reason for this is the fact that Cypress decides the order of execution for test files based on alphabetical order. In this example, we use this naming strategy:
 - `xx_app_spec.js` for the spec files that visit the app and create events
 - `xx_micro_spec.js` for the spec files that query Micro and make the assertions on those events
, where `xx` stands for numbering (but it could be anything as long as it matches uniquely).

```
$ tree testing/cypress/integration

testing/cypress/integration/
├── 01_app_spec.js
├── 01_micro_spec.js
├── 02_app_spec.js
├── 02_micro_spec.js
├── 03_app_spec.js
├── 03_micro_spec.js
```


CHANGELOG
Version 1.0.1 (2021-08-27)
--------------------------
Update Snowplow Micro to 1.2.1 (#24)
Bump django to 3.2.6 (#23)
Upgrade JavaScript Tracker to v3 (#22)
Bump nightwatch to 1.7.8 (#21)
Bump cypress to 8.3.0 (#20)

Version 1.0.0 (2020-11-18)
--------------------------
Update Snowplow Javascript Tracker to v2.16 (#11)
Update Snowplow Micro to v1 (#10)

Version 0.1.1 (2020-10-16)
--------------------------
Add CHANGELOG (#8)
Match chromedriver and chrome version (#6)
Use version number to specify Snowplow Micro Docker image (#4)

Version 0.1.0 (2020-08-06)
--------------------------
Initial release


cypress/jsm/helpers.js
// helper functions for testing with Snowplow Micro


'use strict';


const needParse = [ 'ue_pr', 'co' ];
const needDecode = [ 'ue_px', 'cx' ];


// recurring sanity check
function sane ( n ) {

    if ( n < 0 ) {

        throw( "number of events cannot be negative" );

    }

    return parseInt( n );

}


// Given an array of events and an eventType(string),
// Returns an array of the events having that eventType
function matchByEventType ( eventsArray, eventType ) {

    return eventsArray.filter( hasEventType( eventType ) );

}


// Given an array of events and a schema,
// Returns an array of the events having that schema
function matchBySchema ( eventsArray, schema ) {

    return eventsArray.filter( hasSchema( schema ) );

}


// Given an array of "ue" events and a values object
// Returns an array of the events with matching properties
// [assumes the second argument is an object]
function matchByVals ( eventsArray, valsObj ) {

    return eventsArray.filter( hasValues( valsObj ) );

}


// Given an array of events and a parameters object
// Returns an array of the events with matching parameters
function matchByParams ( eventsArray, paramsObj ) {

    return eventsArray.filter( hasParams( paramsObj ) );

}


// Given an array of events and contexts(array of objects)
// Returns an array of the events with matching contexts
function matchByContexts ( eventsArray, expectedContextsArray ) {

    return eventsArray.filter( hasContexts( expectedContextsArray ) );

}


// ------


function hasEventType ( evType ) {

    return function ( ev ) {

        return ev[ "eventType" ] === evType;

    };

}


function hasSchema ( schema ) {

    return function ( ev ) {

        if ( ev[ "eventType" ] === "ue" ) {

            let ue_pr;

            if ( ev["event"]["parameters"].hasOwnProperty("ue_pr")) {

                ue_pr = JSON.parse( ev[ "event" ][ "parameters" ][ "ue_pr" ] );

            } else {  // must then have ue_px

                let decod_ue_px = base64decode( ev[ "event" ][ "parameters" ][ "ue_px" ] );

                ue_pr = JSON.parse( decod_ue_px );

            }

            return ue_pr[ "data" ][ "schema" ] === schema;

        } else {

            return false;

        }

    };

}


function hasValues ( values ) {

    return function ( ev ) {

        if ( ev[ "eventType" ] === "ue") {

            let ue_pr;

            if ( ev[ "event" ][ "parameters" ].hasOwnProperty( "ue_pr" ) ) {

                ue_pr = JSON.parse( ev[ "event" ][ "parameters" ][ "ue_pr" ] );

            } else {  // must then have ue_px

                let decod_ue_px = base64decode( ev[ "event" ][ "parameters" ][ "ue_px" ] );

                ue_pr = JSON.parse( decod_ue_px );

            }

            let data = ue_pr[ "data" ][ "data" ];

            return Object.keys( values ).every( keyIncludedIn( data ) ) &&
                Object.keys( values ).every( comparesIn( values, data ) );

        } else {

            return false;

        }

    };

}


function hasParams ( expectParams ) {

    return function ( ev ) {

        let actualParams = ev[ "event" ][ "parameters" ];

        let isActualEncoded = needDecode.some( keyIncludedIn( actualParams ) );
        let isExpectedEncoded = needDecode.some( keyIncludedIn( expectParams ) );

        if ( isActualEncoded && !isExpectedEncoded ) {

            actualParams[ "ue_pr" ] = base64decode( actualParams[ "ue_px" ] );
            actualParams[ "co" ] = base64decode( actualParams[ "cx" ] );

        }

        if ( !isActualEncoded && isExpectedEncoded ) {  // that should be rare case

            expectParams[ "ue_pr" ] = base64decode( expectParams[ "ue_px" ] );
            expectParams[ "co" ] = base64decode( expectParams[ "cx" ] );

            // for .every to pass
            delete expectParams[ "ue_px" ];
            delete expectParams[ "cx" ];

        }

        return Object.keys( expectParams ).every( keyIncludedIn( actualParams ) ) &&
            Object.keys( expectParams ).every( comparesIn( expectParams, actualParams ) );

    };

}


function hasContexts ( expCoArr ) {

    return function ( ev ) {

        let p = ev[ "event" ][ "parameters" ];

        if ( p.hasOwnProperty( "co" ) ) {

            let actCoArr = JSON.parse( p[ "co" ] )[ "data" ];

            return compare( expCoArr, actCoArr );

        } else if ( p.hasOwnProperty( "cx" ) ) {

            let co = base64decode( p[ "cx" ] );
            let actCoArr = JSON.parse( co )[ "data" ];

            return compare( expCoArr, actCoArr );

        } else {

            return false;

        }

    };

}


// ------


function keyIncludedIn ( obj ) {

    return function ( key ) {

        return Object.keys( obj ).includes( key );

    };

}


function comparesIn ( expected, actual ) {

    return function ( key ) {

        let expValue, actValue;

        if ( needParse.includes( key ) ) {

            expValue = JSON.parse( expected[ key ] );
            actValue = JSON.parse( actual[ key ] );

        } else {

            expValue = expected[ key ];
            actValue = actual[ key ];

        }

        return compare( expValue, actValue );

    };

}


function compare ( expVal, actVal ) {

    let expType = Object.prototype.toString.call( expVal );
    let actType = Object.prototype.toString.call( actVal );

    if ( expVal === null ) {

        return actVal === null;

    } else if ( expType !== actType ) {

        return false;

    } else if ( expType === "[object Array]" ) {

        return expVal.every( function ( e ) {

            return actVal.some( function ( a ) {

                return compare( e, a );

            });

        });

    } else if ( expType === "[object Object]" ) {

        return Object.keys( expVal ).every( function ( k ) {

            return Object.keys( actVal ).includes( k ) &&
                compare( expVal[ k ], actVal[ k ] );

        });

    } else {

        return expVal === actVal;

    }

}


function base64decode ( encodedData ) {
    //  discuss at: http://locutus.io/php/base64_decode/
    // original by: Tyler Akins (http://rumkin.com)
    // improved by: Thunder.m
    // improved by: Kevin van Zonneveld (http://kvz.io)
    // improved by: Kevin van Zonneveld (http://kvz.io)
    //    input by: Aman Gupta
    //    input by: Brett Zamir (http://brett-zamir.me)
    // bugfixed by: Onno Marsman (https://twitter.com/onnomarsman)
    // bugfixed by: Pellentesque Malesuada
    // bugfixed by: Kevin van Zonneveld (http://kvz.io)
    // improved by: Indigo744
    //   example 1: base64_decode('S2V2aW4gdmFuIFpvbm5ldmVsZA==')
    //   returns 1: 'Kevin van Zonneveld'
    //   example 2: base64_decode('YQ==')
    //   returns 2: 'a'
    //   example 3: base64_decode('4pyTIMOgIGxhIG1vZGU=')
    //   returns 3: '✓ à la mode'

    // decodeUTF8string()
    // Internal function to decode properly UTF8 string
    // Adapted from Solution #1 at https://developer.mozilla.org/en-US/docs/Web/API/WindowBase64/Base64_encoding_and_decoding

    var decodeUTF8string = function ( str ) {
	// Going backwards: from bytestream, to percent-encoding, to original string.
	return decodeURIComponent( str.split('').map( function (c) {

	    return '%' + ('00' + c.charCodeAt( 0 ).toString( 16 ) ).slice( -2 );

	}).join('') );
    };

    var b64 = 'ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789+/=';
    var o1, o2, o3, h1, h2, h3, h4, bits, i = 0,
	ac = 0,
	dec = '',
	tmpArr = [];

    if ( !encodedData ) {

	return encodedData;

    }

    encodedData += '';

    do {
	// unpack four hexets into three octets using index points in b64
	h1 = b64.indexOf( encodedData.charAt( i++ ) );
	h2 = b64.indexOf( encodedData.charAt( i++ ) );
	h3 = b64.indexOf( encodedData.charAt( i++ ) );
	h4 = b64.indexOf( encodedData.charAt( i++ ) );

	bits = h1 << 18 | h2 << 12 | h3 << 6 | h4;

	o1 = bits >> 16 & 0xff;
	o2 = bits >> 8 & 0xff;
	o3 = bits & 0xff;

	if ( h3 === 64 ) {

	    tmpArr[ ac++ ] = String.fromCharCode( o1 );

	} else if ( h4 === 64 ) {

	    tmpArr[ ac++ ] = String.fromCharCode( o1, o2 );

	} else {

	    tmpArr[ ac++ ] = String.fromCharCode( o1, o2, o3 );

	}

    } while ( i < encodedData.length );

    dec = tmpArr.join('');

    return decodeUTF8string( dec.replace(/\0+$/, '') );

}

//
// EXPORTS
//
export {
    sane,
    matchByEventType,
    matchBySchema,
    matchByVals,
    matchByParams,
    matchByContexts,
    compare,
    base64decode
};
